Python revision

msainio opened the pull request:
## Proposed changes

Move content regarding Python package installations, development environments and parallel jobs from Python application page (apps/python.md) to a new Python usage guide (support/tutorials/python-usage-guide.md). Improve and restructure content.

Rework Python application page. Add content like environment package lists, links to official Python documentation and a warning about using the default system Python 3.6.

Restructure and add content to Jupyter application page.

The most substantial changes introduced in this branch are limited to the following pages:
- https://csc-guide-preview.rahtiapp.fi/origin/python-revision/apps/jupyter/
- https://csc-guide-preview.rahtiapp.fi/origin/python-revision/apps/python/
- https://csc-guide-preview.rahtiapp.fi/origin/python-revision/computing/webinterface/jupyter/ (introductory section)
- https://csc-guide-preview.rahtiapp.fi/origin/python-revision/support/tutorials/python-usage-guide/

## Checklist before requesting a review

- [x] I have followed the instructions in the [Contributing](https://github.com/CSCfi/csc-user-guide/blob/master/CONTRIBUTING.md) and [Styleguide](https://github.com/CSCfi/csc-user-guide/blob/master/STYLEGUIDE.md) documents.
- [x] My pull request passes the automated tests. (If not, visit [pull requests at Travis CI](https://app.travis-ci.com/github/CSCfi/csc-user-guide/pull_requests) and have a look at the job log.)
- [x] I have included a link to the appropriate [preview page](https://csc-guide-preview.rahtiapp.fi/origin/) (select your branch from the list).

ktiits commented on docs/apps/python.md:
```diff
@@ -5,256 +5,97 @@ tags:
+module, a
+basic
+[system Python](python.md#system-python)
+is installed by default.
```
```suggestion
is available by default.
```

ktiits commented on docs/support/tutorials/python-usage-guide.md:
````diff
@@ -0,0 +1,254 @@
+	CSC-provided [python-data](../../apps/python-data.md) module:
+
+	```bash
+	cd /projappl/<your_project>  # change this to the appropriate path for your project
````
```suggestion
	cd /projappl/<your_project>/<venv_name>  # change this to the appropriate path for your project
```

mvsjober replied:
Actually I think the original is correct as the command `python3 -m venv --system-site-packages venv` will create the venv directory. Instead that command could be changed to:

`python3 -m venv --system-site-packages <venv_name>`

to make it clear that you can give it whatever name you want.

ktiits replied:
Ok, sorry, if that is the case. But then it could be commented that venv creates this folder. (For example in Tykky empty folder must exist before.)

msainio replied:
Is the project `/projappl` directory the optimal place for creating venvs? At least VSCode automatically sources virtual environments found in an opened "folder", so instructing users to create them in whichever is their working directory might have its benefits

ktiits replied:
/projappl is permanent during projects, from scratch in a half a year, it might get deleted. So in general it is recommended to use /projappl for installations.

msainio replied:
Makes sense :) I changed `venv` to `<venv_name>` and wrote in the info box below that there is no need to create a directory for the environment since `venv` takes care of it

ktiits commented on docs/support/tutorials/python-usage-guide.md:
````diff
@@ -0,0 +1,254 @@
+
+	```bash
+	module load python-data
+	source /projappl/<your_project>/venv/bin/activate
````
```suggestion
	source /projappl/<your_project>/<venv_name>/venv/bin/activate
```

ktiits commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+PC](./remote-dev.md) with some code editors like Visual
+Studio Code.
+
+Finally, several graphical programming environments can be used directly on the
```
Should this be rather first option?

msainio replied:
Yes, it should definitely be recommended over remote dev for most users. It's second now, right after text-based editors ;)

ktiits commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+
+### Jupyter Notebooks
+
+[Jupyter Notebooks](https://jupyter.org/) allows one to run Python code via a web
```
```suggestion
[Jupyter Notebooks](https://jupyter.org/) allows one to run Python code Puhti or Mahti web interface.
```

msainio replied:
I understood that this was referring to how Jupyter notebooks are opened with a web browser when running them locally. In any case, I removed this part and introduce Jupyter more generally as an "interactive programming environment".

mvsjober commented on docs/apps/python.md:
```diff
@@ -5,256 +5,97 @@ tags:
+| [geoconda](geoconda.md) | geoinformatics | [conda](https://a3s.fi/python-pkg-lists/geoconda_3.10.9.yml) |
+| [jax](jax.md) | JAX ML framework | [pip](https://a3s.fi/python-pkg-lists/jax0.4.23_python3.9_cuda12.2_csc_fix1.txt) |
+| [python-data](python-data.md) | data analysis and ML utilities | [conda](https://a3s.fi/python-pkg-lists/python-data-2023-11.yaml) |
+| [pytorch](pytorch.md) | PyTorch ML framework | [pip](https://a3s.fi/python-pkg-lists/pytorch_2.2.1_csc_fix2.txt) |
```
How are these lists generated? And is it automatic? At least for PyTorch it looks like it might just be the requirements file I have used in installation. This is a bit problematic as I use the requirements file only for "simple" installations. It doesn't include packages "manually" installed in the singularity recipe, like for example torch and deepspeed.

msainio replied:
I generated new ones using `list-packages` for the container-based environments and `pip list` for `biopythontools`. I wrote a little script for doing that, maybe it could be set up to update them once a month or so?

mvsjober commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+in a module provided by CSC, don't hesitate to contact our [Service
+Desk](../contact.md).
+
+=== "Using `venv`"
```
As a usability/cosmetic issue I find it confusing with the tab as you can't see what part of the text is the one that will change. Maybe there could be some line or border around the area that belongs to the "tabbable" area?

msainio replied:
Good point, I was wondering about that too. All tabbed content is now inside "default" type admonitions.

mvsjober replied:
Now it much clearer what belongs to the tab. Those boxes look a bit "heavy", but I don't know of a better solution. Maybe ask @joonas-somero ?

joonas-somero replied:
How about a simple horizontal rule? Like this:

```markdown
---

## Creating your own Python environments
```

joonas-somero replied:
Also, should the admonitions have titles? Instead of just `!!! info`, perhaps for example `!!! info "Module compatibility"` would be more appropriate.

joonas-somero replied:
> How about a simple horizontal rule? Like this:

This is what it looks like:
![image](https://github.com/CSCfi/csc-user-guide/assets/50655931/86a0d9a6-83af-4e2b-9049-defb64e593bc)

msainio replied:
@rkronberg suggested removing the nested admonitions, which I did. The boxes look much more visually pleasing to me now. Do we feel there still a need for using something else (e.g. the horizontal rule @joonas-somero suggested)?

![image](https://github.com/CSCfi/csc-user-guide/assets/91546703/7a7f7d0d-c78c-4def-879c-4ff63e3c586e)

joonas-somero replied:
The admonitions are in danger of becoming severely overused and using them merely as a border looks, IMO, really heavy. How about a CSS solution? Here's one with a light border:
![image](https://github.com/CSCfi/csc-user-guide/assets/50655931/44fc660a-6170-4a78-9998-fd3edfa5ec9c)

msainio replied:
Okay, makes sense. I replaced the admonitions with a horizontal rule, since it matches nicely with the automatically created rule at the start of the tab

msainio replied:
I also gave some more descriptive titles to the info boxes

![image](https://github.com/CSCfi/csc-user-guide/assets/91546703/e7395a0d-1359-44ab-831f-05e2727a5e37)

rkronberg replied:
I think this is a good point by @joonas-somero 
Admonitions should be used to draw special attention, and if they are used too much, it seems to defeat the purpose (user is expected to pay special attention to "everything"). In this sense, the horizontal rule would be better. Perhaps custom CSS is overkill?

mvsjober commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+tools](../../data/moving/graphical_transfer.md).
+You can also [edit Python scripts in Puhti from your local
+PC](./remote-dev.md) with some code editors like Visual
+Studio Code.
```
For Visual Studio Code there could be a link to this page: 
https://docs.csc.fi/support/tutorials/remote-dev/#visual-studio-code-with-remote-ssh-plugin
or this one: https://docs.csc.fi/apps/vscode/

msainio replied:
I created a separate subsection for VSCode, since there were already ones for Jupyter and Spyder. It includes links to the VSCode OOD app page as well as the instructions for using the remote ssh plugin

mvsjober commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+[geoconda](../../apps/geoconda.md)
+include the Jupyter notebook package. See the tutorial [how to set up and connect to
+a Jupyter Notebook](./rstudio-or-jupyter-notebooks.md) for using
+Jupyter in CSC environment.
```
If you don't read carefully, on the linked page `rstudio-or-jupyter-notebooks.md` you can easily miss the fact that the easiest (and recommended) way of launching Jupyter Notebook is via Open Ondemand, not via ssh-tunneling etc. We should probably here link directly to the OOD apps.

msainio replied:
The section now links to OOD and ssh-tunneling separately, and OOD is presented as the recommended option.

mvsjober commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,254 @@
+See our GitHub repository for some [examples for using the different parallelization
+options with Puhti](https://github.com/csc-training/geocomputing/tree/master/python/puhti).
+
+The `mpi4py` is not included in the current Python environments in CSC supercomputers,
```
Actually pytorch has mpi4py. Maybe it should be added to python-data in the future?

msainio replied:
The section now mentions that `mpi4py` is included in the `pytorch` module.

msainio commented:
Because of an unfortunate `git pull --rebase`, the commits from `1bd2092` to `c050bed` are duplicates of earlier commits. Apologies for the mess, and thank you @rkronberg for removing the other branches' commits from the history :)

I added `qiskit` to the module list on the Python app page. I also reworked the basic content on Jupyter, but I will not update the instructions on modifying the Jupyter Python environment until the new launch form is ready. I will create a new feature branch for that when it is time.

While I still intend to write some instructions and usage examples for running different kinds of Python jobs, I think it is fine to have those in a different feature branch, too, since I do not really plan to modify the content introduced in this branch. I believe the content updates in this branch form a coherent whole, so from my side, this branch is ready for review.

The most substantial changes introduced in this branch are limited to the following pages:

https://csc-guide-preview.rahtiapp.fi/origin/python-revision/apps/jupyter.md 
https://csc-guide-preview.rahtiapp.fi/origin/python-revision/apps/python.md
https://csc-guide-preview.rahtiapp.fi/origin/python-revision/computing/webinterface/jupyter/ (introductory section)
https://csc-guide-preview.rahtiapp.fi/origin/python-revision/support/tutorials/python-usage-guide/

rkronberg commented:
Perhaps adding a what's new for these very nice improvements would be warranted as well? :)

joonas-somero commented on docs/apps/jupyter.md:
```diff
@@ -1,4 +1,4 @@
----
+Jupyter-
```
This is not on purpose, is it?

msainio replied:
Nope, will fix

joonas-somero commented on docs/apps/by_discipline.md:
```diff
@@ -189,11 +188,12 @@
 * [Jupyter for courses](jupyter-for-courses.md) A version of the Jupyter app for course environments
 * [MySQL](../data/kaivos/overview.md) MySQL database client.
 * [ParaView](paraview.md) Free open-source visualization application
+* [Python](python.md) Programming language and its modules at CSC
```
```suggestion
* [Python](python.md) The programming language and its modules at CSC
```

msainio replied:
It's changed now.

joonas-somero commented on docs/support/tutorials/python-usage-guide.md:
```diff
@@ -0,0 +1,291 @@
+need, but it is missing a few Python packages, you can try installing
+those yourself with the pip package manager.
+
+See the [package lists on our Python application
+page](../../apps/python.md#pre-installed-python-environments) to find out
```
The test scripts in _tests/_ are based on _grep_, so they don't see links that span multiple lines. My suggestion (for now) is to disregard any line length limits when it comes to links.

msainio replied:
All links are now single-line

joonas-somero commented on docs/support/tutorials/python-usage-guide.md:
````diff
@@ -0,0 +1,291 @@
+		python3 -m venv --system-site-packages <venv_name>
+		source <venv_name>/bin/activate
+		pip install whatshap
+		```
+        Unlike for example Tykky, `venv` creates a new directory for the
````
There needs to be an empty line after code blocks. Otherwise the following text will not be placed inside a `<p>` block, but as loose text.

```suggestion
		```

        Unlike for example Tykky, `venv` creates a new directory for the
```

msainio replied:
There is now an empty line after each code block

joonas-somero commented:
Another thing I noticed is that there seems to be a number of tab characters, at least in _python-usage-guide.md_. I'd recommend replacing every tab with four spaces.

msainio commented:
> Another thing I noticed is that there seems to be a number of tab characters, at least in _python-usage-guide.md_. I'd recommend replacing every tab with four spaces.

All tabs in `python-usage-guide.md` and `python.md` replaced with 4 spaces

joonas-somero commented:
@msainio You might want to include `WIP` in the title or convert to a draft PR if you're still editing. If the required statuses pass, someone may prematurely merge this to master. :slightly_smiling_face:

msainio commented:
> @msainio You might want to include `WIP` in the title or convert to a draft PR if you're still editing. If the required statuses pass, someone may prematurely merge this to master. 🙂

Oh, I wasn't changing anything important :) I just figured that since all the reviews had been resolved and the PR was still open I would have a quick look if I spotted any wording that could be marginally improved. I.e. I'm not planning to make any further changes to the content.

rkronberg commented:
I'll go ahead and merge this then! Brilliant work @msainio and everyone :muscle:
